### ISSUE TITLE
vite: module preload is not being handled properly

### ISSUE THREAD

**Issue by lifeiscontent** [remix-run/remix#9037]
### Reproduction

per the docs: https://vitejs.dev/config/build-options#build-modulepreload

> By default, a [module preload polyfill](https://guybedford.com/es-module-preloading-integrity#modulepreload-polyfill) is automatically injected. The polyfill is auto injected into the proxy module of each index.html entry. If the build is configured to use a non-HTML custom entry via build.rollupOptions.input, then it is necessary to manually import the polyfill in your custom entry:

```js
import 'vite/modulepreload-polyfill'
```

### System Info

```shell
System:
    OS: macOS 14.2.1
    CPU: (12) arm64 Apple M2 Max
    Memory: 11.95 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.11.1 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 10.2.4 - /opt/homebrew/bin/npm
    pnpm: 8.15.2 - /opt/homebrew/bin/pnpm
  Browsers:
    Chrome: 122.0.6261.112
    Edge: 122.0.2365.80
    Safari: 17.2.1
```

### Used Package Manager

pnpm

### Expected Behavior

when using vite + remix it automatically injects `vite/modulepreload-polyfill` into its dynamic inputs from rollupOptions

### Actual Behavior

remix doesn't seem to be handling this scenario outlined in the docs, and from what I can tell in Sentry.io our app stability has gone down significantly since switching to vite across different browsers, possible due to this not being implemented.

**Comment by ianserlin** [comment]
@lifeiscontent Not 100% sure, but this may be related to #8803 if you're describing behavior in development, and there may be fixes there that help you.

**Comment by lifeiscontent** [comment]
@ianserlin yeah, that might be it, I'm actually not sure what exactly the fix should look like but the behaviors I'm seeing from upgrading to vite is as follows:

1. when loading a page for the first time some times the loaders fail to run and in the app will show an error boundary (because I throw in the loader if anything goes wrong) so far, this issue only seems to show up in development
2. older browsers like iOS Safari 15.2 don't seem to work anymore because things like Object.hasOwn aren't polyfilled, I've tried using the legacy plugin from vite, but it still doesn't seem to work with remix

**Comment by lifeiscontent** [comment]
@brophdawg11 can we get some eyes on this? the remix vite plugin has been around for awhile now, and I'm still unsure if this is working, specifically when adding `@vitejs/plugin-legacy` to a remix + vite project.

**Comment by brophdawg11** [comment]
Thank you for opening this issue, and our apologies we haven't gotten around to it yet!

With the release of [React Router v7](https://remix.run/blog/react-router-v7) we are sun-setting continued development/maintenance on Remix v2. If you have not already [upgraded to React Router v7](https://reactrouter.com/upgrading/remix), we recommend you do so. We've tried to make the upgrade process as smooth as possible with our [Future Flags](https://remix.run/docs/en/main/start/future-flags). We are now in the process of cleaning up outdated issues and pull requests to improve the overall hygiene of our repositories.

We plan to continue to address 2 types of issues in Remix v2:

- Bugs that pose security concerns
- Bugs that prevent upgrading to React Router v7

If you believe this issue meets one of those criteria, please respond or [create a new issue](https://github.com/remix-run/remix/issues/new?template=bug_report.yml).

For all other issues, ongoing maintenance will be happening in React Router v7, so:

- If this is a bug, please reopen this issue [in that repo](https://github.com/remix-run/react-router/issues/new?template=bug_report.yml) with a new [minimal reproduction](https://reactrouter.com/new) against v7
- If this is a feature request, please open a new [Proposal Discussion](https://github.com/remix-run/react-router/discussions/new?category=proposals) in React Router, and if it gets enough community support it can be considered for implementation

If you have any questions you can always reach out on [Discord](https://rmx.as/discord). Thanks again for providing feedback and helping us make our framework even better!
